### ISSUE TITLE
slightly stricter casting rules for C pointers

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#2057]
IRC user gamester pointed out this example:

```zig
const std = @import("std");

fn foo(b: [*c]i32) void {
    std.debug.warn("Value: {}\n", b.*);
}

pub fn main() void {
    var a: i32 = 40;
    foo(&a); // compiles, runs
    foo(a); // compiles, segfaults
}
```

I created an equivalent C code example:

```c
#include <inttypes.h>
#include <stdio.h>

static void foo(int32_t *b) {
    printf("Value: %d\n", *b);
}

int main() {
    int32_t a = 40;
    foo(a);
    return 0;
}
```

This actually does compile in C (and has the same runtime behavior), but with clang it gives a warning:

```
andy@xps ~/tmp> zig build-exe --c-source test.c --library c
test.c:10:9: warning: incompatible integer to pointer conversion passing
      'int32_t' (aka 'int') to parameter of type 'int32_t *' (aka 'int *'); take
      the address with & [-Wint-conversion]
    foo(a);
        ^
        &
test.c:4:26: note: passing argument to parameter 'b' here
static void foo(int32_t *b) {
                         ^
```

Given that this warning is on by default, I think it would be OK to make this into a compile error. So - definitely an integer other than `usize` should have an error. Potentially even implicit cast to/from `usize` could be a problem if the address wasn't comptime known to be zero.

**Comment by Rocknest** [comment]
Why even cast integer to c-pointer exists? If its used to create null pointers why not use `null`?

**Comment by andrewrk** [comment]
`null` would work once #1967 is implemented. Auto-translated C code will tests such as `ptr == 0` which should continue to work. So comptime known 0 would still be an exception even if this proposal is accepted.

**Comment by Rocknest** [comment]
@andrewrk Is it possible to auto-translate C `ptr != 0` `ptr != NULL` `ptr != (void*)0` as `ptr != null`?

**Comment by andrewrk** [comment]
@Rocknest yes, and that change is now in master branch: https://github.com/ziglang/zig/commit/e3542196c0b0121fb52d9af599b549cf490e47a8

**Comment by digitcrusher** [comment]
Hello, fellow kids. I'm here just to let you know that this issue *still* does *indeed* affect something you'd call "real world users". To make a long story short: I woke up this morning, tried to check the time using my latest and greatest 100% homegrown 70% organic 29.9% Zig clock program I spent the whole weekend working on:
```zig
const std = @import("std");
const c = @cImport(@cInclude("time.h"));

pub fn main() void {
  std.debug.print("{s}", .{c.ctime(c.time(null))});
}
```
And got hit by this *nasty* segmentation fault. It was confusing really. Didn't know what to do. Eventually I went and asked my granddad to show me his date program (He's still the same old C zealot, no matter how many times I point out to him just how rubbish C is 🙄), and then it dawned on me.
```c
#include <stdio.h>
#include <time.h>

int main() {
  time_t t = time(NULL);
  printf("%s", ctime(&t));
}
```
I forgot to put a `&` in front of my `c.time(null)`… Shame on me, I know. But shame on the Zig compiler for not telling me something that it was sooo completely aware of and letting me embarrass myself and my fav language in front of that old fart! The compiler should be *helping* me in our mutual cause of converting all C fans (Ughh, I absolutely detest them) into Zig enjoyers, and not getting in the way of me doing that! Anyways, I recommend that you, Dear Sir Andrew Kelley, shall give your utmost consideration to this pressing issue and not let it fall between the cracks, so that the Greatest Language ever made can be Made Even Greater. You will NOT regret it.
